### PR TITLE
refactor dynamic backend to make use of DLPack

### PR DESCRIPTION
**Pull request by MahmoudAshraf97**
# PR Description
this PR integrates dlpack into the dynamic backend setting process to avoid using numpy, it also fixes minor issues listed below:
1. `ivy.Array.backend` now returns the actual backend of the array, not the current global backend
2. in implicit backend, it's treated as actual numpy backend, you can differentiate between numpy backend and implicit backend by checking the length of `ivy.backend_stack`

### CODE REVIEW

File: ivy/__init__.py [hunk 1]
```diff
@@ -978,7 +978,7 @@ def del_global_attr(attr_name):
     delattr(globals_vars, attr_name)
 
 
-backend = ""
+backend = ivy.utils.backend.handler.implicit_backend
```
Comment (vedpatwardhan) [inline ivy/__init__.py:981]: I don't think we should be doing this, because as a result of this change `ivy.backend` returns `numpy` and so does `ivy.current_backend_str()`. However, if you haven't set a backend and still pass a `torch.Tensor` to an `ivy` function, it would defer to the `torch` backend implementation, which isn't the same as when the `numpy` backend is actually set. 
We should ideally retain the current behaviour when a backend isn't set, what do you think?

File: ivy/data_classes/array/array.py [hunk 2]
```diff
@@ -160,7 +161,7 @@ def _init(self, data, dynamic_backend=None):
         self._dev_str = None
         self._pre_repr = None
         self._post_repr = None
-        self._backend = ivy.backend
+        self._backend = _get_backend_for_arg(self.data.__class__.__module__).backend
```
Comment (vedpatwardhan) [inline ivy/data_classes/array/array.py:164]: I can't really think of a case when an `ivy.Array` is created with a different backend than the one set currently, not sure if that's possible either tbh (as ideally it would throw an error), could you please share an example of the issue being fixed? Thanks!

File: ivy/data_classes/array/array.py [hunk 3]
```diff
@@ -188,28 +189,27 @@ def dynamic_backend(self):
...
-                native_data = _variable_data(self.data)
-                np_data = to_numpy(native_data)
-                new_arr = ivy.array(np_data)
-                self._data = _variable(new_arr).data
+            if ivy_backend.gradients._is_variable(self.data):
+                native_var = ivy_backend.gradients._variable_data(
+                    self,
+                )
+                data = _data_to_new_backend(native_var, ivy_backend).data
+                self._data = _variable(data).data
```
Comment (vedpatwardhan) [inline ivy/data_classes/array/array.py:203]: ig this could also be `ivy_backend.gradients._variable` instead?

File: ivy/functional/backends/paddle/gradients.py [hunk 4]
```diff
@@ -21,8 +21,6 @@
 
 
 def variable(x, /):
-    if ivy.is_int_dtype(x.dtype):
```
Comment (vedpatwardhan) [inline ivy/functional/backends/paddle/gradients.py:24]: I'm not sure why we're changing this, this is actually quite useful to avoid code from breaking when someone incorrectly passes an integer array to execute_with_gradients and misses out on specifying that it should be ignored in `xs_grad_idxs`

File: ivy/functional/ivy/general.py [hunk 5]
```diff
@@ -1826,7 +1826,7 @@ def current_backend_str() -> Union[str, None]:
     """
     fw = current_backend()
     if not backend_stack:
-        return ""
+        return ivy.utils.backend.handler.implicit_backend
```
Comment (vedpatwardhan) [inline ivy/functional/ivy/general.py:1829]: ig we shouldn't make this change for the reason mentioned in the `__init__.py` comment

File: ivy/utils/backend/handler.py [hunk 6]
```diff
@@ -236,9 +236,24 @@ def _handle_backend_specific_vars(target, backend):
         target.set_global_attr("RNG", target.functional.backends.jax.random.RNG)
 
 
-def convert_from_source_backend_to_numpy(variable_ids, numpy_objs, devices):
-    # Dynamic Backend
-    from ivy.functional.ivy.gradients import _is_variable, _variable_data
+def _data_to_new_backend(x, previous_backend):
```
Comment (vedpatwardhan) [inline ivy/utils/backend/handler.py:239]: I'm not sure if I'm testing the wrong example, but if I run the following,
```
import ivy

ivy.set_backend("torch")
x = ivy.array([1., 2., 3.])
ivy.set_backend("tensorflow")
print(ivy.utils.backend.handler._data_to_new_backend(x, ivy.functional.backends.torch))
```
It seems to throw an error `ivy.utils.exceptions.IvyBackendException: tensorflow: dev: 'torch.device' object is not subscriptable`. Probably we could separate out the conversion to and from dlpack in the way we were doing it for conversion to and from numpy in 2 different steps, what do you think?

File: ivy/utils/backend/handler.py [hunk 7]
```diff
@@ -236,9 +236,24 @@ def _handle_backend_specific_vars(target, backend):
...
+    except Exception:
+        np_res = previous_backend.to_numpy(x.data)
+        result = ivy.asarray(np_res, device=device)
+    return result
+
+
+def dynamic_backend_converter(backend_stack):
+    from ivy.functional.ivy.gradients import _variable
+
+    if len(backend_stack) < 2:
```
Comment (vedpatwardhan) [inline ivy/utils/backend/handler.py:253]: I think it's safe to assume that dynamic backend setting would only be used when one backend was set, a few arrays were created and then the user wants to switch to the second backend (i.e. `len(backend_stack)` would always be `>=2`).
Cuz we can't guarantee the backend of the array created when no backend was set (because any type of native array can be passed with implicit backend). And given that implicit backend is only supposed to be used for testing something out quickly, it makes our job much easier when all the arrays created were guaranteed to have the same backend.
Secondly, there might be some issues using a local `ivy` version (from `with_ backend`) and the global one, because the `ivy.Array` in the local version is different from the `ivy.Array` class in the global one, have you tested if the code works as expected?

File: ivy_tests/test_ivy/test_misc/test_backend_utils/test_backend_handler.py [hunk 8]
```diff
@@ -88,32 +88,33 @@ def test_current_backend(backend, array_type):
...
-    "middle_backend,end_backend", [(a, b) for a in backends for b in backends if a != b]
+    "middle_backend,end_backend",
+    [(a, b) for a in backends for b in backends if (a != b and "mxnet" not in [a, b])],
 )
 def test_dynamic_backend_all_combos(middle_backend, end_backend):
     # create an ivy array, container and native container
     a = ivy.array([1, 2, 3])
     b = ivy.array([4, 5, 6])
     ivy_cont = ivy.Container({"w": a, "b": b})
-    nativ_cont = ivy.Container(
```
Comment (vedpatwardhan) [inline ivy_tests/test_ivy/test_misc/test_backend_utils/test_backend_handler.py:98]: are we skipping this test now (with native array containers)?

**Review by vedpatwardhan — CHANGES_REQUESTED** [review]
Hey @MahmoudAshraf97, just requested a few changes, thanks 😄

File: ivy/data_classes/array/array.py (hunk 3 above)
Comment (MahmoudAshraf97) [reply ivy/data_classes/array/array.py:203]: `ivy_backend` is the old backend, but we need to call this function using `ivy` which uses the new backend

File: ivy_tests/test_ivy/test_misc/test_backend_utils/test_backend_handler.py (hunk 8 above)
Comment (MahmoudAshraf97) [reply ivy_tests/test_ivy/test_misc/test_backend_utils/test_backend_handler.py:98]: I gave this a lot of thought, and I couldn't see how the dynamic backend switch should change a non-ivy native variable to the new backend, this is a very invasive behaviour and shouldn't be expected, instead I added a new test case for ivy-wrapped variable container

File: ivy/__init__.py (hunk 1 above)
Comment (MahmoudAshraf97) [reply ivy/__init__.py:981]: we can still differentiate between implicit and numpy backend as I mentioned above by `len(ivy.backend_stack)`, it's demonstrated in the change in `sub_backend_handler.py`, aside from that everything is working as expected
image

File: ivy/data_classes/array/array.py (hunk 2 above)
Comment (MahmoudAshraf97) [reply ivy/data_classes/array/array.py:164]: it's possible to create an array with another backend than the set one using `with_backend`, but you are right, it still returns the correct backend then, reverting this

File: ivy/__init__.py (hunk 1 above)
Comment (vedpatwardhan) [reply ivy/__init__.py:981]: I'm not sure if it will be used as an indicator through, `ivy.backend` being `numpy` would end up being assumed as though the backend set is `numpy`. I think the difference in `ivy.backend` with and without setting the backend is a much clearer indicator that `numpy` would only be used if the input passed to the function doesn't belong to any specific backend (as `current_backend()` returns `numpy` in this case, but the appropriate backend if the input belongs to another backend). I think the previous setup of `numpy` being the "default" backend rather than the "implicit" one is less ambiguous about the usability of `ivy` without setting a backend

File: ivy/data_classes/array/array.py (hunk 2 above)
Comment (vedpatwardhan) [reply ivy/data_classes/array/array.py]: sounds good, thanks!

File: ivy/data_classes/array/array.py (hunk 3 above)
Comment (vedpatwardhan) [reply ivy/data_classes/array/array.py:202]: got it! Thanks for clarifying

File: ivy_tests/test_ivy/test_misc/test_backend_utils/test_backend_handler.py (hunk 8 above)
Comment (vedpatwardhan) [reply ivy_tests/test_ivy/test_misc/test_backend_utils/test_backend_handler.py:98]: yeah sounds good!

File: ivy/__init__.py (hunk 1 above)
Comment (vedpatwardhan) [reply ivy/__init__.py:981]: .

File: ivy/data_classes/array/array.py [hunk 9]
```diff
@@ -401,13 +400,7 @@ def __repr__(self):
                 self._post_repr = ")"
         sig_fig = ivy.array_significant_figures
         dec_vals = ivy.array_decimal_values
-        if self.backend == "" or ivy.is_local():
```
Comment (vedpatwardhan) [inline ivy/data_classes/array/array.py:404]: not sure why we're changing the `__repr__`, ig related to the same point about `numpy` being the implicit backend? This is something which could break if the array was belonging to a different backend than `numpy` right?

File: ivy/functional/backends/paddle/gradients.py (hunk 4 above)
Comment (vedpatwardhan) [reply ivy/functional/backends/paddle/gradients.py:24]: .

**Review by vedpatwardhan — COMMENTED** [review]
Just responded to some conversations and took another look, let's continue on the unresolved conversations, thanks @MahmoudAshraf97 😄
